### PR TITLE
Use iron-list for displaying menu in grapher.

### PR DESCRIPTION
**Pull request by adamcw**
Replaced the `table` in labrad-grapher with an `iron-list` making it substantially more performant even on large datasets.

### CODE REVIEW

File: client-js/app/elements/elements.html [hunk 1]
```diff
@@ -8,6 +8,7 @@
 <link rel="import" href="../bower_components/iron-icons/communication-icons.html">
 <link rel="import" href="../bower_components/iron-icons/iron-icons.html">
 <link rel="import" href="../bower_components/iron-input/iron-input.html">
+<link rel="import" href="../bower_components/iron-list/iron-list.html">
```
Comment (joshm-3b6b58) [inline client-js/app/elements/elements.html]: So this html file is a vestige that we have to get rid of. We're putting imported elements in the file that imports them. Sorry I should make an issue of this I keep forgetting

**Comment by joshm-3b6b58** [comment]
So there's a subtle bug here, that I'm not sure we care about. I'm creating a folder with a large number of entries to test performance, and I notice when a new entry is added the page pops back to the beginning of the list. Perhaps this is a limitation with `<iron-list>`? Maybe there's a different away to append to the list?

Edit: In looking a little deeper in this, this is probably because of the way `listItems` is generated, which I think recreates the list from the arrays every time a new entry is added. Fixing this is outside this PR, I believe.

Edit edit: actually this bug doesn't exist on current master. It stays at the current spot on the list when new entries are added.

**Comment by adamcw** [comment]
This probably does require updating the data binding, I'll look into it and then assess if it seems suitable to be part of this PR.

File: client-js/app/elements/elements.html (hunk 1 above)
Comment (adamcw) [inline client-js/app/elements/elements.html]: Done.

**Comment by joshm-3b6b58** [comment]
Yeah, I've been looking into this a little bit. It looks like scroll to the top on update is the desired behavior for some reason? I can't tell.

**Comment by joshm-3b6b58** [comment]
There are methods where this could be done manually where you can track the `lastVisibleIndex` and update the `scrollToIndex()`

**Comment by adamcw** [comment]
This is now more of a monster of a PR to review than I would like, but I have updated the code to fix the jumping-to-the-top issue.

I completely reworked the data-binding to be incremental, allowing the list to be appended to rather than rewriting the entire thing which was causing iron-list to jump to the top.

I was able to remove the dependency on the kick++ hack which is also nice.

There is probably more refactoring that could go here, such as removing the dirs/datasets variables entirely, but this required more refactoring of `activities.ts` than I was inclined to do in one PR.

File: client-js/app/elements/labrad-grapher.ts [hunk 2]
```diff
@@ -35,30 +52,46 @@ export class LabradGrapher extends polymer.Base {
...
+      this.datasets);
+    this.applyListFilter_();
   }
 
   starClicked() {
     this.showStars = !this.showStars;
     if (this.showStars) this.showTrash = false;
     this.updateButtons();
-    this.kick++;
+    this.applyListFilter_();
```
Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-grapher.ts]: So I'm not familiar with this convention of ending this with underscores, what's the deal with that?

File: client-js/app/elements/labrad-grapher.ts [hunk 3]
```diff
@@ -35,30 +52,46 @@ export class LabradGrapher extends polymer.Base {
...
+
   @listen("keypress")
   onKeyPress(event) {
-    var id = this.selectedId;
+    if (!this.selected) {
+      return "";
+    }
+
+    const selectedIndex = this.listItems.indexOf(this.selected);
+
```
Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-grapher.ts]: Why the whitespace here?

File: client-js/app/elements/labrad-grapher.ts [hunk 4]
```diff
@@ -35,30 +52,46 @@ export class LabradGrapher extends polymer.Base {
...
 
   starClicked() {
     this.showStars = !this.showStars;
     if (this.showStars) this.showTrash = false;
     this.updateButtons();
-    this.kick++;
+    this.applyListFilter_();
   }
 
   trashClicked() {
```
Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-grapher.ts]: I know you didn't name this originally, but `trashClicked()` and `starClicked()` could use more descriptive function names. For a quick read through I can't tell if this is the function that filters the list or tags the datasets.

File: client-js/app/scripts/activities.ts [hunk 5]
```diff
@@ -74,8 +74,8 @@ export class DatavaultActivity implements Activity {
 
   async start(): Promise<ActivityState> {
     console.info('Loading datavault:', this.path);
-    this.api.newDir.add(x => this.onNewDir(), this.lifetime);
-    this.api.newDataset.add(x => this.onNewDataset(), this.lifetime);
+    this.api.newDir.add(x => this.onNewDir(x), this.lifetime);
+    this.api.newDataset.add(x => this.onNewDataset(x), this.lifetime);
```
Comment (joshm-3b6b58) [inline client-js/app/scripts/activities.ts:78]: Why was this change necessary?

Edit: so before `onNewDir()` waited for a new dir to pop up but now `onNewDir` is called with the argument of the new dir when a new dir pops up?

Comment (adamcw) [inline client-js/app/scripts/activities.ts:78]: If you look at 115 and 124 you can see that the new dir and new dataset functions now actually receive the name of the new dataset. This lets them pass just the latest item to the grapher rather than regenerating the whole list by calling the api like it used to.

File: client-js/app/elements/labrad-grapher.ts (hunk 3 above)
Comment (adamcw) [inline client-js/app/elements/labrad-grapher.ts]: I'll remove it.

File: client-js/app/elements/labrad-grapher.ts (hunk 4 above)
Comment (adamcw) [inline client-js/app/elements/labrad-grapher.ts]: There is quite a bit more in this file I wanted to rename/refactor, but I was trying to keep this as small and single-function of a PR as possible.

Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-grapher.ts]: OK good.

File: client-js/app/elements/labrad-grapher.ts (hunk 2 above)
Comment (adamcw) [inline client-js/app/elements/labrad-grapher.ts]: Google style guide for JavaScript (and many languages) has all private members and private methods suffixed with an underscore. I can't seem to find exactly where it is called out, but it is common across the style guides.

https://google.github.io/styleguide/javascriptguide.xml?showone=Visibility__private_and_protected_fields_#Visibility__private_and_protected_fields_

https://google.github.io/styleguide/cppguide.html

I have tried to make new code at least follow the style guide, which leads to a weird mash of styles, but it's somewhat unavoidable without every PR first requiring a PR where the entire file gets refactored which are burdensome to review. If this is the workflow we'd prefer moving forward, we can do that.

**Comment by joshm-3b6b58** [comment]
Very nice. The list renders much faster than before and behaves as I expect when new listings are added.
For 1150 entries this is the performance I see
New: 
image
Old: 
image

LGTM

**Comment by adamcw** [comment]
Does that performance graph include you scrolling to the bottom of the list?

**Comment by joshm-3b6b58** [comment]
This is just initial page load. Scrolling seems fine.

File: client-js/app/elements/labrad-grapher.ts (hunk 2 above)
Comment (maffoo) [inline client-js/app/elements/labrad-grapher.ts]: It's actually recommended _not_ to do this for typescript since we have real visibility modifiers like `private`.

File: client-js/app/elements/labrad-grapher.ts [hunk 6]
```diff
@@ -35,30 +52,46 @@ export class LabradGrapher extends polymer.Base {
...
     this.showTrash = !this.showTrash;
     if (this.showTrash) this.showStars = false;
     this.updateButtons();
-    this.kick++;
+    this.applyListFilter_();
   }
 
-  updateButtons() {
+  updateButtons(): void {
```
Comment (maffoo) [inline client-js/app/elements/labrad-grapher.ts:79]: Is the idea here just to have explicit return type annotations? If we're going to require that on `void` functions, we should be consistent and try to do this everywhere (for example, also on the `onKeyPress` method below). I like the idea of explicit return types on public methods, so I'd be for this.

File: client-js/app/elements/labrad-grapher.ts [hunk 7]
```diff
@@ -35,30 +52,46 @@ export class LabradGrapher extends polymer.Base {
...
   @listen("keypress")
   onKeyPress(event) {
-    var id = this.selectedId;
+    if (!this.selected) {
+      return "";
+    }
+
+    const selectedIndex = this.listItems.indexOf(this.selected);
+
+    var id = this.selected.id;
```
Comment (maffoo) [inline client-js/app/elements/labrad-grapher.ts]: looks like this can also be `const`

File: client-js/app/elements/labrad-grapher.ts [hunk 8]
```diff
@@ -77,6 +110,8 @@ export class LabradGrapher extends polymer.Base {
             tags: ['^star']
           });
         }
+        this.set('listItems.' + selectedIndex + '.starred',
```
Comment (maffoo) [inline client-js/app/elements/labrad-grapher.ts]: typescript has string interpolation:

``` typescript
this.set(`listItems.${selectedIndex}.starred`,
         !this.selected.starred);
```

File: client-js/app/elements/labrad-grapher.ts [hunk 9]
```diff
@@ -134,41 +185,61 @@ export class LabradGrapher extends polymer.Base {
...
+    if (this.filterListItems_([item]).length == 1) {
+      this.push('listItems', item);
     }
-    return items;
+  }
+
+  private filterListItems_(items: Array<ListItem>) {
+    if (this.showTrash) {
+      return items.filter((x) => {
+        return (x.trashed || x.isParent);
```
Comment (maffoo) [inline client-js/app/elements/labrad-grapher.ts]: don't need `return` in arrow methods whose body is a single expression:

``` typescript
return items.filter((x) => (x.trashed || x.isParent));
```

File: client-js/app/elements/labrad-grapher.ts [hunk 10]
```diff
@@ -134,41 +185,61 @@ export class LabradGrapher extends polymer.Base {
...
+
+  private addItemToList_(item: ListItem): void {
+    this.push('listItems_', item);
+    if (this.filterListItems_([item]).length == 1) {
+      this.push('listItems', item);
     }
-    return items;
+  }
+
+  private filterListItems_(items: Array<ListItem>) {
```
Comment (maffoo) [inline client-js/app/elements/labrad-grapher.ts]: Might be better to have a private method that returns the filter function, but does not accept an items array and call filter on it. You would then call `items.filter(this.filterFunction)` instead of `this.filterListItems(items)`, and you could also call `this.filterFunction(item)` instead of `this.filterListItems([item]).length !== 0`.

File: client-js/app/elements/labrad-grapher.ts [hunk 11]
```diff
@@ -134,41 +185,61 @@ export class LabradGrapher extends polymer.Base {
...
   @computed()
-  selectedDataset(selectedId: string): string {
-    if (selectedId.startsWith('dataset:')) {
-      var name = selectedId.substring(8);
+  selectedDataset(selected: {id: string}): string {
+    if (!selected) {
+      return "";
+    }
+
+    if (selected.id.startsWith('dataset:')) {
```
Comment (maffoo) [inline client-js/app/elements/labrad-grapher.ts:240]: I'd suggest pulling out prefixes like `'dataset:'` and `'dir:'` into module-level constants.

File: client-js/app/elements/labrad-grapher.ts [hunk 12]
```diff
@@ -134,41 +185,61 @@ export class LabradGrapher extends polymer.Base {
...
+      });
+    }
+
+    return items.filter((x) => {
+      return !x.trashed || x.isParent;
+    });
+  }
+
+  private applyListFilter_(): void {
+    // Will cause `iron-list` to jump to the top.
```
Comment (maffoo) [inline client-js/app/elements/labrad-grapher.ts]: You mean that the view scrolls to the top of the filtered list? Maybe clarify this comment a bit.

File: client-js/app/elements/labrad-grapher.ts [hunk 13]
```diff
@@ -1,26 +1,43 @@
...
+  @property({type: Array, value: () => { return []; }})
   datasets: Array<{name: string; url: string; tags: Array<string>}>;
 
-  @property({type: String, notify: true})
-  selectedId: string;
+  @property({type: Array, value: () => { return []; }})
+  listItems: Array<ListItem>;
+
+  @property({type: Array, value: () => { return []; }})
+  listItems_: Array<ListItem>;
```
Comment (maffoo) [inline client-js/app/elements/labrad-grapher.ts]: I would suggest changing the names of these properties to something like `filteredListItems` and `allListItems` (or maybe just `listItems` for the latter, since it is clear that it is not filtered).

File: client-js/app/elements/labrad-grapher.ts [hunk 14]
```diff
@@ -95,35 +130,51 @@ export class LabradGrapher extends polymer.Base {
...
+      id: 'dir:' + dir.name,
+      name: dir.name,
+      url: dir.url,
+      isDir: true,
+      isDataset: false,
+      isParent: false,
+      starred: this.isTagged_(dir, 'star'),
+      trashed: this.isTagged_(dir, 'trash')
+    };
+    this.addItemToList_(obj);
```
Comment (maffoo) [inline client-js/app/elements/labrad-grapher.ts]: I would probably drop the declaration and just pass the object literal to the function, but it's not a big deal either way. I'm curious if there's a reason you prefer having the declaration as a separate statement?

``` typescript
this.addItemToList({
  ...
});
```

**Comment by maffoo** [comment]
Arg, sorry, hit wrong button. Just wanted to say please see comments :-)

File: client-js/app/elements/labrad-grapher.ts (hunk 10 above)
Comment (adamcw) [inline client-js/app/elements/labrad-grapher.ts]: I like this idea. Done.

File: client-js/app/elements/labrad-grapher.ts (hunk 13 above)
Comment (adamcw) [inline client-js/app/elements/labrad-grapher.ts]: Agreed, this is very confusing. I was thinking `listItems` and `listItemsView`, but `filteredListItems` sounds a better choice. Done.

File: client-js/app/elements/labrad-grapher.ts (hunk 14 above)
Comment (adamcw) [inline client-js/app/elements/labrad-grapher.ts]: Originally obj was being passed to more than one function, a relic I forgot to reconsider. I agree it would be nicer to just pass the literal directly. Done.

File: client-js/app/elements/labrad-grapher.ts (hunk 11 above)
Comment (adamcw) [inline client-js/app/elements/labrad-grapher.ts:240]: I wasn't sure if this sort of refactoring was appropriate or should be done in a separate PR.

Comment (maffoo) [inline client-js/app/elements/labrad-grapher.ts:240]: Separate PR would be fine; best to keep these small and focused. The state of this code is such that for almost every line that gets touched in a PR, one can see a few other things that need fixing... Glad you're helping fix things up :-)

File: client-js/app/elements/labrad-grapher.ts (hunk 8 above)
Comment (adamcw) [inline client-js/app/elements/labrad-grapher.ts]: Done.

File: client-js/app/elements/labrad-grapher.ts (hunk 9 above)
Comment (adamcw) [inline client-js/app/elements/labrad-grapher.ts]: Done.

File: client-js/app/elements/labrad-grapher.ts (hunk 12 above)
Comment (adamcw) [inline client-js/app/elements/labrad-grapher.ts]: Done.

File: client-js/app/elements/labrad-grapher.ts (hunk 6 above)
Comment (adamcw) [inline client-js/app/elements/labrad-grapher.ts:79]: I think we should aim to have types on everything, private or public, it lets the compiler do a lot of work for us for little effort. Done.

File: client-js/app/elements/labrad-grapher.ts (hunk 7 above)
Comment (adamcw) [inline client-js/app/elements/labrad-grapher.ts]: Done.

**Comment by adamcw** [comment]
A little hard to tell if I've responded to all the comments, but I believe it should be ready for another look.

File: client-js/app/scripts/activities.ts [hunk 15]
```diff
@@ -112,14 +112,22 @@ export class DatavaultActivity implements Activity {
...
-  async onNewDir() {
-    var listing = await this.api.dir(this.path);
-    this.elem.dirs = this.getDirs(listing);
+  private onNewDir(item: {name: string, tags?: Array<string>}) {
+    const dir = {
+      name: item.name,
+      url: this.places.grapherUrl(this.path, item.name),
+      tags: (item.tags) ? item.tags : []
+    };
+    this.elem.newDir(dir);
```
Comment (maffoo) [inline client-js/app/scripts/activities.ts]: Could also omit `const` declarations here and in `onNewDataset` and just pass object literals to the appropriate functions.

**Comment by maffoo** [comment]
One minor comment, otherwise LGTM.

File: client-js/app/scripts/activities.ts (hunk 15 above)
Comment (adamcw) [inline client-js/app/scripts/activities.ts]: Done.

**Comment by adamcw** [comment]
I feel the need to comment while we're on the topic that we may want to avoid `iron-icon` in general unless we actually want to use their data binding abilities and just serve SVG/PNG versions of the icons instead. I will have to do new perf testing given the `iron-list`, but I believe a big chunk of the rendering time is still setting those up. This is sufficient performance for now, but something to be aware of moving forward.
